File: _ETC/B형대비문제/Social media/Social media/정답코드/CPP/slow/3.cpp
//22,276KB	630MS	C++	4,506B
//#include <stdio.h>
//사용자: 본인 id, 팔로잉 정보
//게시물 : 게시물 id 의 발행한 사용자, 게시물id의  좋아요, 게시물의 id의 타임스탬프
//N: 사용자는 1000
// 각 함수 10만회 이하, 단 getfeed 만 5천 이하.
#define MAXN (1000+1)
#define MAXT (100000+1)
 
 
int userInfo[MAXN][MAXN];
typedef struct list {
    int uID;
    int like;
    int timestamp;
}LIST;
 
LIST data[MAXT];
int heap[MAXT];
 
int current_pid;
int heapSize;
 
 
void init(int N)
{
    //printf("N: %d\n", N);
    heapSize = 0;
    for (register int i = 1; i <= N; i++) {
        for (register int j = 1; j <= N; j++) {
            userInfo[i][j] = 0;
        }
        userInfo[i][i] = 1;
    }
}
 
void follow(int uID1, int uID2, int timestamp)
{
    userInfo[uID1][uID2] = 1;
}
 
void makePost(int uID, int pID, int timestamp)
{
    data[pID].uID = uID;
    data[pID].timestamp = timestamp;
    data[pID].like = 0;
    current_pid = pID;
}
 
void like(int pID, int timestamp)
{
    data[pID].like++;
}
 
/* 오리지날
int heapPush(int value)
{
    if (heapSize + 1 > MAX_SIZE)
    {
        printf("queue is full!");
        return 0;
    }
 
    heap[heapSize] = value;
 
    int current = heapSize;
    while (current > 0 && heap[current] < heap[(current - 1) / 2])
    {
        int temp = heap[(current - 1) / 2];
        heap[(current - 1) / 2] = heap[current];
        heap[current] = temp;
        current = (current - 1) / 2;
    }
 
    heapSize = heapSize + 1;
 
    return 1;
}
*/
 
int compare(LIST a, LIST b)
{
    int ret = 0;
 
    if (a.like > b.like) {
        ret = 1;
    }
    else if (a.like == b.like) {
        if (a.timestamp > b.timestamp) {
            ret = 1;
        }
        else {
            ret = 0;
        }
    }
    else {
        ret = 0;
    }
 
    return ret;
}
 
int heapPush(int value)
{
    if (heapSize > MAXT)
    {
        //printf("queue is full!");
        return 0;
    }
 
    heap[heapSize] = value;
 
 
    int current = heapSize;
 
    //compare(                data[heap[current]], data[heap[(current - 1) / 2]]                       )
 
 
    while (current > 0 && compare(data[heap[current]], data[heap[(current - 1) / 2]]))
    {
 
        int temp = heap[(current - 1) / 2];
        heap[(current - 1) / 2] = heap[current];
        heap[current] = temp;
        current = (current - 1) / 2;
    }
 
    heapSize = heapSize + 1;
 
    return 1;
}
 
 
 
int heapPop(int* value)
{
    if (heapSize <= 0)
    {
        return -1;
    }
 
    *value = heap[0];
    heapSize = heapSize - 1;
 
    heap[0] = heap[heapSize];
 
    int current = 0;
    while (current * 2 + 1 < heapSize)
    {
        int child;
        if (current * 2 + 2 == heapSize)
        {
            child = current * 2 + 1;
        }
        else
        {
            child = compare(data[heap[current * 2 + 1]], data[heap[current * 2 + 2]]) ? current * 2 + 1 : current * 2 + 2;
        }
 
        if (compare(data[heap[current]], data[heap[child]]))
        {
            break;
        }
 
        int temp = heap[current];
        heap[current] = heap[child];
        heap[child] = temp;
 
        current = child;
    }
    return 1;
}
 
/*
1. 게시된 지 1,000 초 이내인 게시글이 그렇지 않은 게시글보다 우선순위가 높다.
2. 게시된 지 1,000 초 이내인 게시글들 중에는 “like” 가 많은 게시글의 우선순위가 높다.
3. 게시된 지 1,000 초 이내이면서 “like” 의 개수가 같은 게시글들 중에는 “timestamp” 가 높은 게시글의 우선순위가 높다.
4. 게시된 지 1,000 초를 초과한 게시글의 경우, “timestamp” 가 높은 게시글의 우선순위가 높다.
*/
 
void getFeed(int uID, int timestamp, int pIDList[])
{
    //printf("getfeed uid:%d timestamp:%d\n", uID, timestamp);
    int temp = 0;
    int temp2[MAXT] = { 0, };
    int over1000_Cnt = 0;
    //int candidate[MAXT] = { 0, };
    for (register int i = current_pid; i >= 1; i--) {
        if (userInfo[uID][data[i].uID] == 0) continue;
 
        //if (userInfo[uID][data[i].uID] == 1) {
        /*if (data[i].uID == uID || userInfo[uID][data[i].uID] == 1) {*/
        int delta = timestamp - data[i].timestamp;
        if (delta <= 1000) {
 
            /*
            for (int m = 0; m < 10; m++) {
                if (data[i].like > data[candidate[m]].like) {
                    for (int n = 8; n >= m; n--) {
                        candidate[n + 1] = candidate[n];
                    }
                    candidate[m] = i;
                    break;
                }
            }
            */
            //printf("data[%d].like: %d, timestampe:%d\n", i, data[i].like, data[i].timestamp);
 
            heapPush(i);
 
 
        }
        else {
            if (over1000_Cnt == 10) break;
            temp2[over1000_Cnt++] = i;//
        }
        //}
    }
 
 
 
    //int cnt = 0;
    //for (int i = 0; i < 10; i++) {
    //  if (candidate[i] != 0) {
    //      pIDList[i] = candidate[i];
    //      cnt++;
    //  }
    //}
 
 
    int cnt = 0;
 
    for (int i = 0; i < 10; i++) {
        int value = 0;
        heapPop(&value);
        if (value != 0) {
            pIDList[i] = value;
            /*printf("heap result pIDList[%d]:%d\n", i, pIDList[i]);*/
            cnt++;
        }
    }
 
    //for (int i = 0; i < 10; i++) {
    //  printf("temp2[%d]: %d\n", i, temp2[i]);
    //}
 
 
    for (int i = 0; i < 10; i++) {
        if (cnt < 10) {
            pIDList[cnt++] = temp2[i];
        }
    }
 
    heapSize = 0;
}

File: 3.ETC/B형대비문제/Social media/Social media/정답코드/Java/mid/3.java
//108,320KB	486MS	JAVA	6,536B
class UserSolution {
    int PCNT = 10;
    int MAX_SIZE = 100001;
    int P_MAX_SIZE;
    int runCount;
    class User {
        int uID = 1;
        FList fList = new FList();
        PList pList = new PList();
        User prev;
    }
     
    class Post {
        int pID;
        int like;
        int timestamp;
        Post prev;
        Post() {
             
        }
        Post(int id, int time) {
            pID = id;
            timestamp = time;
        }
    }
    User uList[];
    Post pList[];
    static int baseTime;
     
     
    public void init(int N)
    {
         P_MAX_SIZE = N +1;
         uList = new User[P_MAX_SIZE];
         pList = new Post[MAX_SIZE];
         baseTime = 0;
         runCount = 0;
          
    }
 
    public void follow(int uID1, int uID2, int timestamp)
    {
        //System.out.println("run count = " + runCount);
        runCount++;
        if (uList[uID1] == null) {
            uList[uID1] = new User();
        }
        if (uList[uID2] == null) {
            uList[uID2] = new User();
        }
        uList[uID1].fList.add(uList[uID2]);
    }
 
    public void makePost(int uID, int pID, int timestamp) {
        //System.out.println("run count = " + runCount);
        runCount++;
        if (uList[uID] == null) {
            uList[uID] = new User();
        }
        Post p = new Post(pID, timestamp);
        pList[pID] = p;
        uList[uID].pList.add(p);
 
    }
 
    public void like(int pID, int timestamp)
    {
        //System.out.println("run count = " + runCount);
        runCount++;
        pList[pID].like++;
    }
 
    public void getFeed(int uID, int timestamp, int pIDList[])
    {
        //System.out.println("run count = " + runCount);
         
        runCount++;
        if (runCount == 84) {
            //System.out.println("###");
        }
 
        baseTime = timestamp;
 
         
        int pSize = uList[uID].pList.size();
        int fSize = uList[uID].fList.getFeedSize();
        heapInit(pSize + fSize);
        uList[uID].pList.getFeed();
        uList[uID].fList.getFeed();
        if (runCount == 84) {
            //System.out.println("");
        }
 
        Post post[] = new Post[PCNT];
        for (int i = 0; i < PCNT; i++) {
            post[i] = heapPop();
            if (post[i] != null) {
                pIDList[i] = post[i].pID;
                //System.out.print(pIDList[i] + " ");
            }
        }
        //System.out.println("");
    }
     
    public static boolean checkpriority (Post a, Post b) {  //a가 크면 true
        int timeA = baseTime - a.timestamp;
        int timeB = baseTime - b.timestamp;
        if (timeA <= 1000  && timeB <= 1000 ) {  // 둘다 1000 미만이면
            if (a.like > b.like)
                return true;
            else if (a.like == b.like) {
                return timeA < timeB ? true : false;
            } else {
                return false;
            }
        } else {    //1000초 이상이 있으면 시간순으로
            return timeA < timeB ? true : false; 
        }
    }
 
    Post heap[];
    int heapSize = 0;
 
    void heapInit(int size)
    {
        heapSize = 0;
        heap = new Post[size];
    }
 
    void heapPush(Post value)
    {
        if (heapSize + 1 > MAX_SIZE)
        {
            return;
        }
 
        heap[heapSize] = value;
 
        int current = heapSize;
        while (current > 0 && checkpriority(heap[current], heap[(current - 1) / 2])) 
        {
            Post temp = heap[(current - 1) / 2];
            heap[(current - 1) / 2] = heap[current];
            heap[current] = temp;
            current = (current - 1) / 2;
        }
 
        heapSize = heapSize + 1;
    }
 
    Post heapPop()
    {
        if (heapSize <= 0)
        {
            return null;
        }
 
        Post value = heap[0];
        heapSize = heapSize - 1;
 
        heap[0] = heap[heapSize];
 
        int current = 0;
        while (current < heapSize && current * 2 + 1 < heapSize)
        {
            int child;
            if (current * 2 + 2 >= heapSize)
            {
                child = current * 2 + 1;
            }
            else
            {
                child = checkpriority(heap[current * 2 + 1], heap[current * 2 + 2]) ? current * 2 + 1 : current * 2 + 2;
            }
 
            if (checkpriority(heap[current], heap[child]))
            {
                break;
            }
 
            Post temp = heap[current];
            heap[current] = heap[child];
            heap[child] = temp;
 
            current = child;
        }
        return value;
    }
 
    private class PostNode {
        Post item;
        PostNode next;
        PostNode prev;
 
        PostNode(PostNode prev, Post element, PostNode next) {
            this.item = element;
            this.next = next;
            this.prev = prev;
        }
    }
     
    class PList{
        int size = 0;
        PostNode last;
        PostNode first;
        int modCount = 0;
         
        void add(Post e) {
            final PostNode l = last;
            final PostNode newNode = new PostNode(l, e, null);
            last = newNode;
            if (l == null)
                first = newNode;
            else
                l.next = newNode;
            size++;
            modCount++;
        }
         
        PostNode node(int index) {
            if (index < (size >> 1)) {
                PostNode x = first;
                for (int i = 0; i < index; i++)
                    x = x.next;
                return x;
            } else {
                PostNode x = last;
                for (int i = size - 1; i > index; i--)
                    x = x.prev;
                return x;
            }
        }
         
        public Post get(int index) {
            return node(index).item;
        }
         
        public int size() {
            return size;
        }
         
        public void getFeed() {
            PostNode x = first;
            for (int i = 0; i < size; i++) {
                heapPush(x.item);
                x = x.next;
            }
        }
    } 
     
     
     
    private class UserNode {
        User item;
        UserNode next;
        UserNode prev;
 
        UserNode(UserNode prev, User element, UserNode next) {
            this.item = element;
            this.next = next;
            this.prev = prev;
        }
    }
     
    class FList{
        int size = 0;
        UserNode last;
        UserNode first;
        int modCount = 0;
         
        void add(User e) {
            final UserNode l = last;
            final UserNode newNode = new UserNode(l, e, null);
            last = newNode;
            if (l == null)
                first = newNode;
            else
                l.next = newNode;
            size++;
            modCount++;
        }
         
        UserNode node(int index) {
            if (index < (size >> 1)) {
                UserNode x = first;
                for (int i = 0; i < index; i++)
                    x = x.next;
                return x;
            } else {
                UserNode x = last;
                for (int i = size - 1; i > index; i--)
                    x = x.prev;
                return x;
            }
        }
         
        public User get(int index) {
            return node(index).item;
        }
         
        public int size() {
            return size;
        }
         
        public int getFeedSize() {
            int feedSize = 0;
            UserNode x = first;
            for (int i = 0; i < size; i++) {
                feedSize += x.item.pList.size();
                x = x.next;
            }
            return feedSize;
        }
         
        public void getFeed() {
             
            UserNode x = first;
            for (int i = 0; i < size; i++) {
                x.item.pList.getFeed();
                x = x.next;
            }
        }
 
    } 
   
}